crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -250,26 +324,224 @@ bitflags::bitflags! {
+    pub transform: Affine3,
+    /// Four 16-bit unsigned normalized UV values packed into a [`UVec2`]:
+    ///
+    /// ```text
+    ///                         <--- MSB                   LSB --->
+    ///                         +---- min v ----+ +---- min u ----+
+    ///     lightmap_uv_rect.x: vvvvvvvv vvvvvvvv uuuuuuuu uuuuuuuu,
+    ///                         +---- max v ----+ +---- max u ----+
+    ///     lightmap_uv_rect.y: VVVVVVVV VVVVVVVV UUUUUUUU UUUUUUUU,
+    ///
+    /// (MSB: most significant bit; LSB: least significant bit.)
+    /// ```
+    pub lightmap_uv_rect: UVec2,
+    /// Various flags.
+    pub mesh_flags: MeshFlags,
+}
+
+impl RenderMeshInstanceShared {
+    fn from_components(
+        previous_transform: Option<&PreviousGlobalTransform>,
+        handle: &Handle<Mesh>,
+        not_shadow_caster: bool,
+        no_automatic_batching: bool,
+    ) -> Self {
+        let mut mesh_instance_flags = RenderMeshInstanceFlags::empty();
+        mesh_instance_flags.set(RenderMeshInstanceFlags::SHADOW_CASTER, !not_shadow_caster);
+        mesh_instance_flags.set(
+            RenderMeshInstanceFlags::AUTOMATIC_BATCHING,
+            !no_automatic_batching,
+        );
+        mesh_instance_flags.set(
+            RenderMeshInstanceFlags::HAVE_PREVIOUS_TRANSFORM,
+            previous_transform.is_some(),
+        );
+
+        RenderMeshInstanceShared {
+            mesh_asset_id: handle.id(),
+
+            flags: mesh_instance_flags,
+            material_bind_group_id: AtomicMaterialBindGroupId::default(),
+        }
+    }
+
+    /// Returns true if this entity is eligible to participate in automatic
+    /// batching.
     pub fn should_batch(&self) -> bool {
-        self.automatic_batching && self.material_bind_group_id.get().is_some()
+        self.flags
+            .contains(RenderMeshInstanceFlags::AUTOMATIC_BATCHING)
+            && self.material_bind_group_id.get().is_some()
+    }
+}
+
+/// Information that the render world keeps about each entity that contains a
+/// mesh.
+///
+/// The set of information needed is different depending on whether CPU or GPU
+/// [`MeshUniform`] building is in use.
+#[derive(Resource)]
+pub enum RenderMeshInstances {
+    /// Information needed when using CPU mesh uniform building.
+    CpuBuilding(EntityHashMap<RenderMeshInstanceCpu>),
+    /// Information needed when using GPU mesh uniform building.
+    GpuBuilding(EntityHashMap<RenderMeshInstanceGpu>),
+}
+
+impl RenderMeshInstances {
+    fn new(use_gpu_instance_buffer_builder: bool) -> RenderMeshInstances {
+        if use_gpu_instance_buffer_builder {
+            RenderMeshInstances::GpuBuilding(EntityHashMap::default())
+        } else {
+            RenderMeshInstances::CpuBuilding(EntityHashMap::default())
+        }
     }
+
+    /// Returns the ID of the mesh asset attached to the given entity, if any.
+    pub(crate) fn mesh_asset_id(&self, entity: Entity) -> Option<AssetId<Mesh>> {
+        match *self {
```

Comment:
So performance of these relies on CPU branch prediction working well and always going the same way, which it will. Right?

Reply:
The three places we use this are:
* `extract_lightmaps`
* `SetMeshBindGroup::render`
* `DrawMesh::render`

I've never seen `extract_lightmaps` show up in profiles. The other two can show up there, but I don't see any reasonable way to specialize them for CPU and GPU versions. So I don't think we have any other choice here.

Reply:
In any case, I'll add specific versions of the methods if you know you have a `RenderMeshInstancesCpu` or `RenderMeshInstancesGpu`. That way in the future maybe we can amortize the cost.

Reply:
`queue_material_meshes` and similar methods are now split into two, one for CPU mesh instance data generation and one for GPU mesh instance data generation. If you ask me, I prefer just eating the cost of the branch, as the code is cleaner, but I'll defer to you.